packages/component/src/component_container.ts

```diff
@@ -0,0 +1,67 @@
+/**
+ * @license
+ * Copyright 2019 Google Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { Provider } from './provider';
+import { Component } from './component';
+
+export class ComponentContainer {
+  private providers = new Map<string, Provider>();
+
+  constructor(private name: string) {}
+
+  /**
+   *
+   * @param component Component being added
+   * @param overwrite When a component with the same name has already been registered,
```

Comment:
What about two methods, like `addComponent` and `overwriteComponent`?

Reply:
I usually don't like flag arguments, since it's harder to tell what it's doing when you're reading code that calls this function (or you have to add comments like `addComponent(component, /* overWrite */ true);`).

I guess it's okay if it's mostly for testing, but TBH I'd rather not use the same app instance across tests :smile:. Up to you.

Reply:
Yeah, flag arguments are not ideal. Added a new method `addOrOverwriteComponent()`.

`database` tests share `app` a lot, but I don't feel like changing it at the moment. Also there is a request from `@firebase/testing` to be able to overwrite `auth-internal` component inside an initialized app that has the default `auth-internal`.

Reply:
Fair enough, that sounds like a lot of work.